backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/interceptor/C_Invoice.java

```diff
@@ -37,32 +46,46 @@
 import org.compiere.util.Env;
 import org.springframework.stereotype.Component;
 
+import static de.metas.contracts.modular.ComputingMethodType.AVERAGE_CONTRACT_SPECIFIC_PRICE_METHODS;
 import static de.metas.contracts.modular.ModelAction.COMPLETED;
 import static de.metas.contracts.modular.ModelAction.REVERSED;
 
 @Component
+@RequiredArgsConstructor
 @Interceptor(I_C_Invoice.class)
 public class C_Invoice
 {
-	private final IInvoiceBL invoiceBL = Services.get(IInvoiceBL.class);
+	@NonNull private final IInvoiceBL invoiceBL = Services.get(IInvoiceBL.class);
+	@NonNull private final IFlatrateBL flatrateBL = Services.get(IFlatrateBL.class);
 
-	private final ModularContractService contractService;
-
-	public C_Invoice(@NonNull final ModularContractService contractService)
-	{
-		this.contractService = contractService;
-	}
+	@NonNull private final ModularContractService contractService;
+	@NonNull private final ModularContractLogService modularContractLogService;
+	@NonNull private final ModularContractSettingsService modularContractSettingsService;
+	@NonNull private final ModularContractLogHandlerRegistry logHandlerRegistry;
 
 	@DocValidate(timings = ModelValidator.TIMING_AFTER_COMPLETE)
 	public void afterComplete(@NonNull final I_C_Invoice invoiceRecord)
 	{
 		invokeHandlerForEachLine(invoiceRecord, COMPLETED);
 	}
 
-	@DocValidate(timings = { ModelValidator.TIMING_AFTER_REVERSEACCRUAL, ModelValidator.TIMING_AFTER_REVERSECORRECT })
+	@DocValidate(timings = { ModelValidator.TIMING_AFTER_REVERSEACCRUAL, ModelValidator.TIMING_AFTER_REVERSECORRECT, ModelValidator.TIMING_AFTER_VOID })
 	public void afterReverse(@NonNull final I_C_Invoice invoiceRecord)
 	{
 		invokeHandlerForEachLine(invoiceRecord, REVERSED);
+		final InvoiceId invoiceId = InvoiceId.ofRepoId(invoiceRecord.getC_Invoice_ID());
```

Comment:
Instead of bringing all the services and writing the logic here, I advise you to create for handling the invoice reversal in one of the services (e.g. ModularContractService)  and just call it here.

Reply:
Will be hard to move this to an service without creating circular bean creations, especially `logHandlerRegistry` won't work

Reply:
done, except logHandlerRegistry

![image](https://github.com/user-attachments/assets/d2dce98c-b090-4882-b3dd-396ae68387e0)